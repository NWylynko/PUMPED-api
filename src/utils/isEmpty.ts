export const isArrayEmpty = (array: any[]): true | undefined => (array.length === 0 ? true : undefined);

export const isEmpty = (object: any): true | undefined => (!object ? true : undefined);
