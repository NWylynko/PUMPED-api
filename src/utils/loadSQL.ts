import fs from 'fs';
import path from 'path';
import sqlParser from 'sql-template-tag';

const loadSQL = (file: string) => {
  const sql = fs.readFileSync(path.resolve(__dirname, '../../SQL/', `${file}.SQL`), 'utf-8');

  return sqlParser([sql]).sql;
};

export default loadSQL;
