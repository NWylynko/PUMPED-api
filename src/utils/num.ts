const num = (n: string) => parseInt(n, 10);

export default num;
