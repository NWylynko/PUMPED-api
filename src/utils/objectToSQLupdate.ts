import { escape } from 'sqlstring';

const objectToSQLupdate = (fields: any) => {
  let sql = '';

  Object.keys(fields).forEach((partName, index, array) => {
    // if the current item is the length of the array then its the last item
    // the last item cant have a comma
    const comma = index === array.length - 1 ? '' : ',';

    sql += `${escape(partName)} = ${escape(fields[partName])}${comma} `;
  });

  return sql;
};

export default objectToSQLupdate;
