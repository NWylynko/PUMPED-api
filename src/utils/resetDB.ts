import { promises as fs } from 'fs';
import db, { SQL } from '../db';

export const createTables = () => Promise.all([
  db.run(SQL.tables.Brand),
  db.run(SQL.tables.Collection),
  db.run(SQL.tables.Colour),
  db.run(SQL.tables.ColourImage),
  db.run(SQL.tables.Customer),
  db.run(SQL.tables.Image),
  db.run(SQL.tables.Order),
  db.run(SQL.tables.OrderItem),
  db.run(SQL.tables.Review),
  db.run(SQL.tables.Section),
  db.run(SQL.tables.Shoe),
  db.run(SQL.tables.ShoeTag),
  db.run(SQL.tables.Stock),
  db.run(SQL.tables.Style),
  db.run(SQL.tables.Tag),
  db.run(SQL.tables.WishList),
]);

const resetDB = async () => {
  await db.close();
  await fs.unlink(`./main.${process.env.NODE_ENV}.db`);
  await db.open();
  await createTables();
};

export const addTestData = () => db.exec(SQL.testData);

export const resetDBForTest = async () => {
  await db.close();
  await db.open();
  await createTables();
  await db.exec(SQL.testData);
};

export default resetDB;
