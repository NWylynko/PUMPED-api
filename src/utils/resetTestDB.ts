import { promises as fs } from 'fs';
import db from '../db';

const reset = async () => {
  await db.close();
  await fs.unlink('./main.test.db');
  await fs.copyFile('./main.testData.db', './main.test.db');
  await db.open();
};

export default reset;
