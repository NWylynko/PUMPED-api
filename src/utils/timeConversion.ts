function timeConversion(milliseconds: number) {
  // adapted from https://stackoverflow.com/a/32180863

  const seconds = (milliseconds / 1000);
  const minutes = (milliseconds / (1000 * 60));
  const hours = (milliseconds / (1000 * 60 * 60));
  const days = (milliseconds / (1000 * 60 * 60 * 24));

  if (seconds < 60) {
    return `${Math.floor(seconds)} Seconds`;
  } if (minutes < 60) {
    return `${Math.floor(minutes)} Minutes`;
  } if (hours < 24) {
    return `${Math.floor(hours)} Hours`;
  }
  return `${Math.floor(days)} Days`;
}

export default timeConversion;
