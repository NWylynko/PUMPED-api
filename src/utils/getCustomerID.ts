import { Request } from 'express';
import num from './num';

const getCustomerID = (req: Request): number => {
  const CustomerID = req.header('CustomerID');

  if (!CustomerID) {
    throw new Error('no customer id supplied');
  }

  return num(CustomerID);
};

export default getCustomerID;
