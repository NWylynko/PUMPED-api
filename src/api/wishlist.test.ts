// eslint-disable-next-line import/no-extraneous-dependencies
import supertest from 'supertest';
import express from '../app';
import { resetDBForTest } from '../utils/resetDB';

const app = supertest(express);

beforeEach(async (done) => { await resetDBForTest(); done(); });

describe('GET wishlist', () => {
  [1, 2, 3].forEach(id => {
    it(`get shoes for customer ${id}`, async () => {
      const response = await app.get(`/wishlist/${id}`);
  
      const json = JSON.parse(response.text);
  
      expect(json).toMatchSnapshot();
    })
  })

})

describe('POST wishlist', () => {
  [1, 2, 3].forEach(CustomerID => {
    [1, 2, 3].forEach(ShoeID => {
      it(`add shoe (${ShoeID}) to customer (${CustomerID})`, async () => {
        const response = await app
          .post(`/wishlist/${CustomerID}`)
          .type('json')
          .send(JSON.stringify({
            ShoeID
          }));
    
        const json = JSON.parse(response.text);
    
        expect(json).toMatchSnapshot();

        const getWishlist = await app
          .get(encodeURI(`/wishlist/${CustomerID}`));
    
        const getWishlistJson = JSON.parse(getWishlist.text);
    
        expect(getWishlistJson).toMatchSnapshot();
      })
    })
  })

})