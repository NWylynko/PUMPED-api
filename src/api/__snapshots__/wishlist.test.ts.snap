// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET wishlist get shoes for customer 1 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`GET wishlist get shoes for customer 2 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET wishlist get shoes for customer 3 1`] = `
Object {
  "data": Array [],
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (1) 1`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "ShoeID": "1",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (1) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (2) 1`] = `
Object {
  "data": Object {
    "CustomerID": "2",
    "ShoeID": "1",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (2) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (3) 1`] = `
Object {
  "data": Object {
    "CustomerID": "3",
    "ShoeID": "1",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (1) to customer (3) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (1) 1`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "ShoeID": "2",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (1) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (2) 1`] = `
Object {
  "data": Object {
    "CustomerID": "2",
    "ShoeID": "2",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (2) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (3) 1`] = `
Object {
  "data": Object {
    "CustomerID": "3",
    "ShoeID": "2",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (2) to customer (3) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (1) 1`] = `
Object {
  "data": Object {
    "CustomerID": "1",
    "ShoeID": "3",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (1) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 3,
      "Section": "womens",
      "Stars": 0,
      "Style": "runners",
      "description": "sprint",
      "name": "kicks",
      "price": 120,
      "releaseDate": 2098475234,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (2) 1`] = `
Object {
  "data": Object {
    "CustomerID": "2",
    "ShoeID": "3",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (2) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 3,
      "Section": "womens",
      "Stars": 0,
      "Style": "runners",
      "description": "sprint",
      "name": "kicks",
      "price": 120,
      "releaseDate": 2098475234,
    },
  ],
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (3) 1`] = `
Object {
  "data": Object {
    "CustomerID": "3",
    "ShoeID": "3",
  },
  "success": true,
}
`;

exports[`POST wishlist add shoe (3) to customer (3) 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 3,
      "Section": "womens",
      "Stars": 0,
      "Style": "runners",
      "description": "sprint",
      "name": "kicks",
      "price": 120,
      "releaseDate": 2098475234,
    },
  ],
  "success": true,
}
`;
